### ISSUE TITLE
XHR request after preflight doesn't send FormData body

### ISSUE THREAD

**Issue by BehindTheMath** [jsdom/jsdom#2300]
When using XHR to send a request that has a `FormData` body and requires a preflight, the `FormData` body is not sent with the actual request.

### Basic info:

- **Node.js version:** 8.9.3
- **jsdom version:** 11.5.1

### Minimal reproduction case

```js
const jsdom = require("jsdom");
const {window} = new jsdom.JSDOM();
const {FormData, XMLHttpRequest} = window;

const formData = new FormData();
formData.append("input", "value");

const xhr = new XMLHttpRequest();

// This URL sends back the request data
const url = "https://httpbin.org/post";

xhr.onreadystatechange = function() {
    if (xhr.readyState === 4 && xhr.status === 200) {
        callback(xhr.responseText, xhr);
    }
};

xhr.open("POST", url, true);

// Forces a preflight to be sent
xhr.setRequestHeader("X-Some-Header", "some-data");

xhr.send(formData);

function callback(responseText) {
    const response = JSON.parse(responseText);
    console.dir("Response:", response);
    console.log("Form data sent:", response.form);
    console.log("True if no form-data was sent:", Object.keys(response.form).length === 0)
}

```

[Here's](https://repl.it/@BehindTheMath/KookyRustyEmacs) a runnable example of the above code.

**Expected behavior**: Running the above code should return the data from the FormData object, which would indicate that it was sent correctly.
**Actual behavior**: The response contains no form data, showing no body was sent.

### How does similar code behave in browsers?

[Here's](https://codepen.io/BehindTheMath/pen/WKjwNJ?editors=1011) a codepen demonstrating parallel code using the browser's XHR implementation. This works correctly, and the response is as expected.

### Cause

I believe the cause is due to a race condition in Jsdom's code that occurs when a request with FormData requires a preflight request. When [`XMLHttpRequest.send()`](https://github.com/jsdom/jsdom/blob/d6f8a97b5fb7709d6ad0215c1ae95fd4cab58489/lib/jsdom/living/xmlhttprequest.js#L529) is called, the following steps are performed:

1. An XHR client is created. ([`xmlhttprequest.js`](<https://github.com/jsdom/jsdom/blob/d6f8a97b5fb7709d6ad0215c1ae95fd4cab58489/lib/jsdom/living/xmlhttprequest.js#L664>), [`xhr-utils.js`](<https://github.com/jsdom/jsdom/blob/d6f8a97b5fb7709d6ad0215c1ae95fd4cab58489/lib/jsdom/living/xhr-utils.js#L146>))
2. The XHR client [checks](<https://github.com/jsdom/jsdom/blob/d6f8a97b5fb7709d6ad0215c1ae95fd4cab58489/lib/jsdom/living/xhr-utils.js#L335>) if a preflight request is necessary, and if it is, it is built and sent, passing a callback to get the response. ([`xhr-utils.js`](<https://github.com/jsdom/jsdom/blob/d6f8a97b5fb7709d6ad0215c1ae95fd4cab58489/lib/jsdom/living/xhr-utils.js#L335-L412>))
3. Since the request is async, execution continues. The XHR client object is returned to `xmlhttprequest.js`, and right before returning to the caller, `flag.body` and `flag.formData` are cleared. ([`xmlhttprequest.js`](<https://github.com/jsdom/jsdom/blob/d6f8a97b5fb7709d6ad0215c1ae95fd4cab58489/lib/jsdom/living/xmlhttprequest.js#L738-L739>))
4. When the preflight response returns and the callback is executed, the response is checked to make sure it's valid. If it is, the actual request is built and sent. ([`xhr-utils.js`](https://github.com/jsdom/jsdom/blob/d6f8a97b5fb7709d6ad0215c1ae95fd4cab58489/lib/jsdom/living/xhr-utils.js#L384), [`xhr-utils.js`](https://github.com/jsdom/jsdom/blob/d6f8a97b5fb7709d6ad0215c1ae95fd4cab58489/lib/jsdom/living/xhr-utils.js#L311))

This is where the race condition comes in. When the request is built in step 4, the XHR client checks the local `flag.formData` to see if the body is a FormData instance, and if it is, it adds that data to the `request` instance. However, the local `flag.formData` references the same value as `flag.formData` of `xmlhttprequest.js`, which was already cleared in step 3. So the FormData is never sent, since by that time jsdom thinks it doesn't exist.

If the line adding a custom header is commented out, a preflight request is no longer needed. In that case, step 4 is performed synchronously after step 2, so `flag.formData` is still true, and it is sent as expected.

I don't understand why step 3 is done at all. I'm also not sure why this isn't an issue when sending a regular body.

**Comment by BehindTheMath** [comment]
@domenic Is there something I'm missing here?

**Comment by Zirro** [comment]
@BehindTheMath Thanks for the report! Since Domenic is quite busy (we all work on jsdom in our spare time), finding or writing a [Web Platform Test](https://github.com/jsdom/jsdom/blob/master/Contributing.md#web-platform-feature-tests) that passes in browsers but fails in jsdom would be the quickest way to get a potential fix merged.

**Comment by Zirro** [comment]
In most cases it shouldn't require the most recent version of the WPT repo in order to test a new WPT against jsdom. Adding the test files temporarily inside `test/web-platform-tests/tests/xhr` and running `yarn test-wpt` (or `yarn test-wpt --fgrep "xhr/formdata-with-preflight.htm" to test that specific file) should be enough.

That said, I've submitted [a pull request](https://github.com/jsdom/jsdom/pull/2365) with the changes needed for the latest WPTs if you'd like to use that instead for testing.

**Comment by BehindTheMath** [comment]
>Adding the test files temporarily inside test/web-platform-tests/tests/xhr and running yarn test-wpt (or `yarn test-wpt --fgrep "xhr/formdata-with-preflight.htm" to test that specific file) should be enough.

I'm still getting an error: `Expectation pattern "formdata-with-preflight.htm" does not match any test files`.

I see that it's not in `test\web-platform-tests\wpt-manifest.json`, even after I rebuilt it.

**Comment by wongjiahau** [comment]
Can anyone help me? I'm still facing this problem with Jest even though I upgraded jsdom to the latest version.

**Comment by Behinder** [comment]
Same story here but with jQuery Ajax and express js :(
